File: mobile/lib/agent/scene_user_actions.dart
import 'scene_bindings.dart';

/// User-confirmed deletion: detach linked items; never erase another character.
Map<String, dynamic> sceneEntityRemoval(Map<String, dynamic> scene, String id) {
  final entities = (scene['entities'] as List).cast<Map>();
  final entity = entities.where((e) => e['id'] == id).firstOrNull;
  if (entity == null) throw StateError('SCENE_STALE');
  if (entity['kind'] == 'character' &&
      entities.where((e) => e['kind'] == 'character').length == 1) {
    throw StateError('SCENE_LAST_CHARACTER');
  }
  final affected =
      entities.where((e) => e['ownerId'] == id || e['wearerId'] == id).toList();
  final facts =
      (scene['facts'] as List).where((f) => f['entityId'] == id).toList();
  final relations = (scene['relations'] as List)
      .where((r) => r['actorId'] == id || r['targetId'] == id)
      .toList();
  bool parentLocked(dynamic id) =>
      entities.any((e) => e['id'] == id && e['locked'] == true);
  if ([entity, ...affected].any((e) =>
          e['locked'] == true ||
          parentLocked(e['ownerId']) ||
          parentLocked(e['wearerId'])) ||
      facts.any((f) => f['locked'] == true) ||
      relations.any((r) =>
          r['locked'] == true ||
          parentLocked(r['actorId']) ||
          parentLocked(r['targetId']))) {
    throw StateError('SCENE_LOCKED');
  }
  final patch = {
    'revision': scene['revision'],
    'operations': [
      {'collection': 'entities', 'id': id, 'before': entity, 'after': null},
      for (final e in affected)
        {
          'collection': 'entities',
          'id': e['id'],
          'before': e,
          'after': {
            for (final entry in e.entries)
              if (!(['ownerId', 'wearerId'].contains(entry.key) &&
                  entry.value == id))
                entry.key: entry.value,
          }
        },
      for (final f in facts)
        {'collection': 'facts', 'id': f['id'], 'before': f, 'after': null},
      for (final r in relations)
        {'collection': 'relations', 'id': r['id'], 'before': r, 'after': null},
    ]
  };
  return {
    'patch': patch,
    'next': applyScenePatch(scene, patch, byUser: true),
    'affected': affected,
    'facts': facts,
    'relations': relations
  };
}

File: src/tavern/scene-user-actions.ts
import { applyScenePatch, type SceneBindings, type SceneOperation } from './scene-bindings';

/** Delete only the selected entity. Other people's belongings stay, with obsolete links removed. */
export function sceneEntityRemoval(scene: SceneBindings, id: string) {
  const entity = scene.entities.find(e => e.id === id);
  if (!entity) throw Error('SCENE_STALE');
  if (entity.kind === 'character' && scene.entities.filter(e => e.kind === 'character').length === 1) throw Error('SCENE_LAST_CHARACTER');
  const affected = scene.entities.filter(e => e.ownerId === id || e.wearerId === id);
  const facts = scene.facts.filter(f => f.entityId === id);
  const relations = scene.relations.filter(r => r.actorId === id || r.targetId === id);
  const parentLocked = (entityId: string | undefined) => scene.entities.some(e => e.id === entityId && e.locked);
  if (entity.locked || parentLocked(entity.ownerId) || parentLocked(entity.wearerId)
    || affected.some(e => e.locked || parentLocked(e.ownerId) || parentLocked(e.wearerId))
    || facts.some(f => f.locked) || relations.some(r => r.locked || parentLocked(r.actorId) || parentLocked(r.targetId))) throw Error('SCENE_LOCKED');
  const operations: SceneOperation[] = [
    { collection: 'entities', id, before: entity, after: null },
    ...affected.map(e => {
      const after = { ...e };
      if (after.ownerId === id) delete after.ownerId;
      if (after.wearerId === id) delete after.wearerId;
      return { collection: 'entities' as const, id: e.id, before: e, after };
    }),
    ...facts.map(f => ({ collection: 'facts' as const, id: f.id, before: f, after: null })),
    ...relations.map(r => ({ collection: 'relations' as const, id: r.id, before: r, after: null })),
  ];
  return { patch: { revision: scene.revision, operations }, affected, facts, relations,
    next: applyScenePatch(scene, { revision: scene.revision, operations }, true) };
}
